volttron/platform/agent/base_historian.py

```diff
@@ -1518,41 +1520,54 @@ def remove_successfully_published(self, successful_publishes,
         If None is found in `successful_publishes` we assume that everything
         was published.
 
-        :param successful_publishes: List of records that was published.
+        :param successful_publishes: Set of records that was published.
         :param submit_size: Number of things requested from previous call to
                             :py:meth:`get_outstanding_to_publish`
 
-        :type successful_publishes: list
+        :type successful_publishes: set
         :type submit_size: int
 
         """
 
         #_log.debug("Cleaning up successfully published values.")
         c = self._connection.cursor()
-
-        if None in successful_publishes:
-            c.execute('''DELETE FROM outstanding
-                        WHERE ROWID IN
-                        (SELECT ROWID FROM outstanding
-                          ORDER BY ts LIMIT ?)''', (submit_size,))
-            if self._record_count < c.rowcount:
-                self._record_count = 0
+        try:
+            if None in successful_publishes:
+                if not self._dupe_ids:
+                    c.execute('''DELETE FROM outstanding
+                                WHERE ROWID IN
+                                (SELECT ROWID FROM outstanding ORDER BY ts LIMIT ?)''', (submit_size,))
+                    if self._record_count < c.rowcount:
+                        self._record_count = 0
+                    else:
+                        self._record_count -= c.rowcount
+                else:
+                    # this is a rare case, so adding logs to help track and debug when this case does occur
```

Comment:
we could always delete based on unique_ids unless the first query would perform better. i am not sure if that is true though.

Reply:
How would the query look like instead? I though I was deleting on unique_ids  using the following:

```sql
'''DELETE FROM outstanding 
  WHERE id = ?''',
((_id,) for _id in self._unique_ids)
```

Reply:
I meant you could use the above query in both if and else clause. If there are no duplicates won't unique ids simply contain all the ids the subquery (SELECT ROWID FROM outstanding ORDER BY ts LIMIT ?)''', (submit_size,)) return ?

Reply:
I see your point. I initially did it this way to highlight and isolate the case of duplicates, maintain the supposed performance gain on executing one query in the case of no duplicates, and not delete current production code. But you are right, in both cases, the query will delete items whether it has duplicates or not.

Reply:
I did a quick and dirty performance test between execute and executemany on a DELETE query with a size of 1000 records. It appears that the difference is negligible:

`executemany`
PASSED [100%]deleted
elapsed: 0.0029566287994384766

`execute`
PASSED [100%]deleted
elapsed: 0.002218961715698242

Reply:
Based on offline conversation with @schandrika, I've flattened the logic to be a simple if/else.